org/mozilla/jss/crypto/KBKDFParameterSpec.java

```diff
@@ -0,0 +1,115 @@
+package org.mozilla.jss.crypto;
+
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.KeySpec;
+
+import java.util.Arrays;
+import java.util.ArrayList;
+
+import org.mozilla.jss.util.NativeEnclosure;
+
+public abstract class KBKDFParameterSpec extends NativeEnclosure implements AlgorithmParameterSpec, KeySpec {
+    protected PKCS11Algorithm prf;
+    protected KBKDFDataParameter[] params;
+    protected KBKDFDerivedKey[] additional_keys;
```

Comment:
Yeah... This is where I say that the native layer is tough. So what's happening is these are stored on the class itself, right? In C pseudo-code, we want to eventually do:

```c
CK_PRF_DATA_PARAM data_params = calloc(this.params.length, sizeof(CK_PRF_DATA_PARAM)); 
for (size_t index = 0; index < this.params.length; index++) {
    // Copy data from the KBKDFDataParameter class to the actual C struct. 
    data_params[index] = *this.params[index].ptr;
}

CK_SP800_108_KDF_PARAMS kdf_params = { this.params.length, data_params,  ... }

call_this_kdf(kdf_params);
``` 

(Where `this` refers to the `KBKDFParameterSpec` Java class instance and `this.params` is shorthand for actually accessing that array from the JNI layer. It isn't that simple sadly). 

By using `NativeEnclosure`, each class/parameter/key/... worries about how it gets serialized in a vacuum. It allocates memory for itself, it frees its own memory, it transfers whatever data it needs to between the allocated memory and the Java class.

We can then blindly copy the contents of / use those pointers from the JNI C layer (after accessing the value of the pointer) -- which makes that part really easy. 

Anyhow, using `ArrayList` makes the whole "accessing" part harder and requires jumping through more hoops to access the data. Definitely possible, just more work. I'd prefer to simplify the JNI layer as much as possible, hence a straight array.

--- 

And to answer your question, "eh", probably not. It'd probably be fine to only add a 'setParams' / `setAdditionalKeys` method and not an `addParameter` method. But I don't know how it'll get used in the end, so...

(We also don't need any additional keys so that'll largely go unused from our portion -- its just parameters that we care about for SCP03).